pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidCallingFinalizeRule.java

```diff
@@ -4,68 +4,100 @@
+    }
+
+    private boolean isFinalizeMethodDeclaration(ASTMethodDeclaration methodDeclaration) {
+        return "finalize".equals(methodDeclaration.getName()) && methodDeclaration.getArity() == 0;
+    }
+
+    private boolean callsNotSuperFinalizeMethod(ASTBlock block) {
+        List<ASTPrimaryExpression> primaryExpressions = block.findDescendantsOfType(ASTPrimaryExpression.class);
+        for (ASTPrimaryExpression primaryExpression : primaryExpressions) {
+            if (isNotSuperFinalizeMethodCall(primaryExpression)) {
+                return true;
+            }
         }
-        addViolation(ctx, pp);
-        return super.visit(pp, ctx);
+        return false;
     }
 
-    private boolean checkForViolation(ScopedNode node) {
-        MethodScope meth = node.getScope().getEnclosingScope(MethodScope.class);
-        if (meth != null && "finalize".equals(meth.getName())) {
-            return false;
+    private boolean isNotSuperFinalizeMethodCall(ASTPrimaryExpression primaryExpression) {
+        return isFinalizeMethodCall(primaryExpression) && isNotSuperMethodCall(primaryExpression);
+    }
+
+    private boolean isNotSuperMethodCall(ASTPrimaryExpression primaryExpression) {
+        ASTPrimaryPrefix primaryPrefix = primaryExpression.getFirstChildOfType(ASTPrimaryPrefix.class);
+        return primaryPrefix == null || !primaryPrefix.usesSuperModifier();
+    }
+
+    private boolean callsFinalizeMethod(ASTBlock block) {
+        List<ASTPrimaryExpression> primaryExpressions = block.findDescendantsOfType(ASTPrimaryExpression.class);
+        for (ASTPrimaryExpression primaryExpression : primaryExpressions) {
+            if (isFinalizeMethodCall(primaryExpression)) {
+                return true;
+            }
         }
-        if (meth != null && checked.contains(meth)) {
-            return false;
+        return false;
+    }
+
+    private boolean isFinalizeMethodCall(ASTPrimaryExpression primaryExpression) {
+        return hasFinalizeName(primaryExpression) && getArgsCount(primaryExpression) == 0;
+    }
+
+    private boolean hasFinalizeName(ASTPrimaryExpression primaryExpression) {
+        List<JavaNode> expressionNodes = primaryExpression.findDescendantsOfType(JavaNode.class);
+        for (JavaNode expressionNode : expressionNodes) {
+            if (isFinalizeName(expressionNode.getImage())) {
+                return true;
+            }
         }
-        if (meth != null) {
-            checked.add(meth);
+        return false;
+    }
+
+    private boolean isFinalizeName(String name) {
+        return name != null && FINALIZE_METHOD_PATTERN.matcher(name).find();
+    }
+
+    private int getArgsCount(ASTPrimaryExpression primaryExpression) {
+        ASTPrimarySuffix primarySuffix = primaryExpression.getFirstChildOfType(ASTPrimarySuffix.class);
+        if (primarySuffix != null) {
+            int argsCount = primarySuffix.getArgumentCount();
+            return max(argsCount, 0);
```

Comment:
I'm unsure this is actually right… `getArgumentCount` returns -1 if the suffix is not actually a method call if I recall correctly… this would interpret it as an actual no-arg invocation…